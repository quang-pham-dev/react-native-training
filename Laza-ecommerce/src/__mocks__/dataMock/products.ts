import { ProductProps } from 'types/Products';

export const product: ProductProps = {
  id: '0',
  brandId: '1',
  name: 'Nike',
  price: 100,
  type: 'shoes',
  like: true,
  title: 'Nike Sportswear Club Fleece',
  source:
    'https://images.unsplash.com/photo-1542291026-7eec264c27ff?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470&q=80',
  description:
    "Nike is a multinaelichick and retired quarterback Jack Nicklaus. Nike's products include apparel, shoes, equipment, and accessories. Nike also owns and operates the majority of the world's sporting goods brands, including Adidas, Puma, and Fila.",
  reviewer: {
    id: '2',
    name: 'John Doe',
    date: '13, Sep, 2020',
    image:
      'https://images.unsplash.com/photo-1609505848912-b7c3b8b4beda?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=765&q=80',
  },
  imageReview: [
    {
      id: '0',
      image:
        'https://images.unsplash.com/photo-1499852848443-3004d6dc4cfc?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1374&q=80',
    },
    {
      id: '1',
      image:
        'https://images.unsplash.com/photo-1618898909019-010e4e234c55?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=687&q=80',
    },
    {
      id: '2',
      image:
        'https://images.unsplash.com/photo-1575537302964-96cd47c06b1b?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470',
    },
    {
      id: '3',
      image:
        'https://images.unsplash.com/photo-1518656306295-aa28b28b2504?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470',
    },
    {
      id: '4',
      image:
        'https://images.unsplash.com/photo-1609505848912-b7c3b8b4beda?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=765&q=80',
    },
    {
      id: '5',
      image:
        'https://images.unsplash.com/photo-1575537302964-96cd47c06b1b?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470',
    },
  ],
  rating: '4.5',
  comment:
    'lorem ipsum dolor sit amet consectetur adipisicing elit. Quisquam,adipisicing elit. Quisquam, quia.',
};

export const products: ProductProps[] = [
  {
    id: '0',
    brandId: '1',
    name: 'Nike',
    price: 100,
    type: 'shoes',
    like: true,
    title: 'Nike Sportswear Club Fleece',
    source:
      'https://images.unsplash.com/photo-1542291026-7eec264c27ff?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470&q=80',
    description:
      "Nike is a multinaelichick and retired quarterback Jack Nicklaus. Nike's products include apparel, shoes, equipment, and accessories. Nike also owns and operates the majority of the world's sporting goods brands, including Adidas, Puma, and Fila.",
    reviewer: {
      id: '2',
      name: 'John Doe',
      date: '13, Sep, 2020',
      image:
        'https://images.unsplash.com/photo-1609505848912-b7c3b8b4beda?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=765&q=80',
    },
    imageReview: [
      {
        id: '0',
        image:
          'https://images.unsplash.com/photo-1499852848443-3004d6dc4cfc?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1374&q=80',
      },
      {
        id: '1',
        image:
          'https://images.unsplash.com/photo-1618898909019-010e4e234c55?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=687&q=80',
      },
      {
        id: '2',
        image:
          'https://images.unsplash.com/photo-1575537302964-96cd47c06b1b?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470',
      },
      {
        id: '3',
        image:
          'https://images.unsplash.com/photo-1518656306295-aa28b28b2504?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470',
      },
      {
        id: '4',
        image:
          'https://images.unsplash.com/photo-1609505848912-b7c3b8b4beda?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=765&q=80',
      },
      {
        id: '5',
        image:
          'https://images.unsplash.com/photo-1575537302964-96cd47c06b1b?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470',
      },
    ],
    rating: '4.5',
    comment:
      'lorem ipsum dolor sit amet consectetur adipisicing elit. Quisquam,adipisicing elit. Quisquam, quia.',
  },
  {
    id: '0',
    brandId: '1',
    name: 'Nike',
    price: 100,
    type: 'shoes',
    like: true,
    title: 'Nike Sportswear Club Fleece',
    source:
      'https://images.unsplash.com/photo-1542291026-7eec264c27ff?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470&q=80',
    description:
      "Nike is a multinaelichick and retired quarterback Jack Nicklaus. Nike's products include apparel, shoes, equipment, and accessories. Nike also owns and operates the majority of the world's sporting goods brands, including Adidas, Puma, and Fila.",
    reviewer: {
      id: '2',
      name: 'John Doe',
      date: '13, Sep, 2020',
      image:
        'https://images.unsplash.com/photo-1609505848912-b7c3b8b4beda?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=765&q=80',
    },
    imageReview: [
      {
        id: '0',
        image:
          'https://images.unsplash.com/photo-1499852848443-3004d6dc4cfc?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1374&q=80',
      },
      {
        id: '1',
        image:
          'https://images.unsplash.com/photo-1618898909019-010e4e234c55?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=687&q=80',
      },
      {
        id: '2',
        image:
          'https://images.unsplash.com/photo-1575537302964-96cd47c06b1b?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470',
      },
      {
        id: '3',
        image:
          'https://images.unsplash.com/photo-1518656306295-aa28b28b2504?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470',
      },
      {
        id: '4',
        image:
          'https://images.unsplash.com/photo-1609505848912-b7c3b8b4beda?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=765&q=80',
      },
      {
        id: '5',
        image:
          'https://images.unsplash.com/photo-1575537302964-96cd47c06b1b?crop=entropy&cs=tinysrgb&fm=jpg&ixlib=rb-1.2.1&q=80&raw_url=true&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470',
      },
    ],
    rating: '4.5',
    comment:
      'lorem ipsum dolor sit amet consectetur adipisicing elit. Quisquam,adipisicing elit. Quisquam, quia.',
  },
];
