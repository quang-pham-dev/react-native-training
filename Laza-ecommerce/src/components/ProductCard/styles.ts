import { StyleSheet } from 'react-native';

// Themes
import Colors from 'themes/Colors';
import Fonts from 'themes/Fonts';
import Metrics from 'themes/Metrics';

const styles = StyleSheet.create({
  productCardContainer: {
    width: 165
  },

  imageWrapper: {
    width: '100%',
    borderRadius: 15,
    overflow: 'hidden',
    backgroundColor: Colors.secondaryBackground
  },

  image: {
    width: '100%',
    height: 203,
    alignItems: 'flex-end',
    resizeMode: 'contain'
  },

  iconHeartWrapper: {
    position: 'absolute',
    top: 0,
    right: 0,
    zIndex: 1
  },

  iconHeart: {
    width: 16,
    height: 15,
    resizeMode: 'contain',
    marginTop: 17,
    marginRight: 14
  },

  productTitle: {
    fontSize: Fonts.size.min,
    fontFamily: Fonts.fontFamily.Inter_500Medium,
    color: Colors.textBlack,
    marginTop: Metrics.margin.xsm
  },

  productType: {
    fontSize: Fonts.size.min,
    fontFamily: Fonts.fontFamily.Inter_500Medium,
    color: Colors.textBlack,
    marginBottom: Metrics.margin.xsm
  },

  price: {
    fontSize: Fonts.size.small,
    fontFamily: Fonts.fontFamily.Inter_600SemiBold,
    color: Colors.textBlack,
    marginTop: Metrics.margin.xsm
  }
});

export default styles;
