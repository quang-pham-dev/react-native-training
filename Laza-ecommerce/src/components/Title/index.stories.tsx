import React, { useState } from 'react';
import { storiesOf } from '@storybook/react-native';
import Title from './index';
import CenterView from '../../../storybook/stories/CenterView';

storiesOf('Title', module)
  .addDecorator(getStory => <CenterView>{getStory()}</CenterView>)
  .add('normal', () => <Title titleName='title' />)
  .add('headingPage', () => <Title.HeadingPage titleName='Welcome' />)
  .add('subHeadingPage', () => (
    <Title.subHeadingPage titleName='Please enter your data to continue' />
  ))
  .add('titleLeftSection', () => <Title.titleLeftSection titleName='New Product' />)
  .add('titleRightSection', () => <Title.titleRightSection titleName='View all' />);
