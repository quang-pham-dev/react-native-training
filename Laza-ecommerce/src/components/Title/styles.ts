import { StyleSheet } from 'react-native';

// Themes
import Colors from 'themes/Colors';

const styles = StyleSheet.create({
  title: {
    textTransform: 'none',
    color: Colors.black,
  },
});

export default styles;
