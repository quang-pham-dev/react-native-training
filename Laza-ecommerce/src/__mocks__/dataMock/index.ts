import { user } from './user';
import { brand, brands } from './brands';
import { product, products } from './products';

export { brand, brands, product, products, user };
