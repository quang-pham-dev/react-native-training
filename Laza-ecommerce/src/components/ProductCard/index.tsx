import React, { memo, useCallback } from 'react';
import { Image, ImageBackground, Pressable, Text, View } from 'react-native';

// LIBS
import isEqual from 'react-fast-compare';

// Components
import Title from 'components/Title';

// Types
import { IProductCardProps } from 'types/models/Products';

// Themes
import IMAGES from 'themes/Images';

// Styles
import styles from './styles';

const ProductCard = ({
  product,
  onPressProductCard,
  onPressLikeProduct,
  productCardStyles
}: IProductCardProps) => {
  // handle action press product card with id
  const handlePressCardProduct = useCallback(() => {
    onPressProductCard(product.id);
  }, [product.id, onPressProductCard]);

  // handle action press like product
  const handlePressLikeProduct = useCallback(() => {
    onPressLikeProduct(product);
  }, [product, onPressLikeProduct]);

  return (
    <Pressable
      testID='productCard'
      onPress={handlePressCardProduct}
      style={[styles.productCardContainer, productCardStyles]}
    >
      <View style={styles.imageWrapper}>
        <ImageBackground style={styles.image} source={{ uri: product?.source }} />
        <Pressable onPress={handlePressLikeProduct} style={styles.iconHeartWrapper}>
          <Image
            style={styles.iconHeart}
            source={product?.like ? IMAGES.iconHeartLiked : IMAGES.iconHeart}
          />
        </Pressable>
      </View>
      <Title titleStyles={styles.productTitle} titleName={product?.title} />
      <Title titleStyles={styles.productType} titleName={product?.type} />
      <Text style={styles.price}>{`$ ${product?.price}`}</Text>
    </Pressable>
  );
};

export default memo(ProductCard, isEqual);
