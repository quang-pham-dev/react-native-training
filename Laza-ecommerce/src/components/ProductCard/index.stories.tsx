import React from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react-native';
import CenterView from '../../../storybook/stories/CenterView';
import { product } from '__mocks__/dataMock/products';
import ProductCard from './index';

storiesOf('Product Card', module)
  .addDecorator(getStory => <CenterView>{getStory()}</CenterView>)
  .add('Default', () => (
    <ProductCard
      product={product}
      onPressProductCard={action('clicked')}
      onPressLikeProduct={action('clicked')}
    />
  ));
