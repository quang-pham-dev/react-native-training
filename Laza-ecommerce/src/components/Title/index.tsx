import React, { memo } from 'react';
import { Text } from 'react-native';

// LIBS
import isEqual from 'react-fast-compare';

// Types
import { ITitleProps } from 'types/components/Title';

// Styles
import styles from './styles';

const Title = ({ titleName, titleStyles }: ITitleProps) => {
  return <Text style={[styles.title, titleStyles]}>{titleName}</Text>;
};

export default memo(Title, isEqual);
