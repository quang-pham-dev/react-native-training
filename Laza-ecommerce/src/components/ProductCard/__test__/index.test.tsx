import React from 'react';
import { Image, ImageBackground, Pressable, Text } from 'react-native';

// LIBS
import renderer from 'react-test-renderer';

// Components
import ProductCard from 'components/ProductCard';
import Title from 'components/Title';

// Mocks
import { product } from '__mocks__/dataMock/products';

// Utils
import { navigationMock } from 'utils/testMock';

describe('Product Card Component', () => {
  const props = {
    product,
    onPressProductCard: navigationMock.navigate,
    onPressLikeProduct: jest.fn(),
  };

  const tree = renderer.create(<ProductCard {...props} />);
  test('Should render correctly', () => {
    const component = tree.toJSON();
    expect(component).toMatchSnapshot();
  });

  test('Should render with ImageBackground', () => {
    const image = tree.root.findAllByType(ImageBackground)[0];
    expect(image.props.source.uri).toBe(product.source);
  });

  test('Should render with title', async () => {
    const title = tree.root.findAllByType(Title)[0];
    expect(title.props.titleName).toEqual(product.title);
    expect(title.props.titleName).toBeTruthy();
  });

  test('Should render with type', async () => {
    const type = tree.root.findAllByType(Title)[1];
    expect(type.props.titleName).toEqual(product.type);
    expect(type.props.titleName).toBeTruthy();
  });

  test('Should render with price', async () => {
    const price = tree.root.findAllByType(Text)[2];
    expect(price.props.children).toEqual(`$ ${product.price}`);
    expect(price.props.children).toBeTruthy();
  });

  test('should call function handlePressCardProduct', () => {
    const press = tree.root.findAllByType(Pressable)[0];
    press.props.onPress();
    expect(navigationMock.navigate).toHaveBeenCalled();
  });
  test('should call function handlePressLikeProduct', () => {
    const press = tree.root.findAllByType(Pressable)[1];
    press.props.onPress();
    expect(props.onPressLikeProduct).toHaveBeenCalled();
  });
});
