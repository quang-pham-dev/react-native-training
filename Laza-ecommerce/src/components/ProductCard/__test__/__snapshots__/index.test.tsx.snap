// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card Component Should render correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "width": 165,
      },
      undefined,
    ]
  }
  testID="productCard"
>
  <View
    style={
      Object {
        "backgroundColor": "#F5F6FA",
        "borderRadius": 15,
        "overflow": "hidden",
        "width": "100%",
      }
    }
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        Object {
          "alignItems": "flex-end",
          "height": 203,
          "resizeMode": "contain",
          "width": "100%",
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "https://images.unsplash.com/photo-1542291026-7eec264c27ff?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1470&q=80",
          }
        }
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "height": 203,
              "width": "100%",
            },
            undefined,
          ]
        }
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "position": "absolute",
          "right": 0,
          "top": 0,
          "zIndex": 1,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "height": 15,
            "marginRight": 14,
            "marginTop": 17,
            "resizeMode": "contain",
            "width": 16,
          }
        }
      />
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "color": "#000",
          "textTransform": "none",
        },
        Object {
          "color": "#1D1E20",
          "fontFamily": "Inter_500Medium",
          "fontSize": 11,
          "marginTop": 5,
        },
      ]
    }
  >
    Nike Sportswear Club Fleece
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#000",
          "textTransform": "none",
        },
        Object {
          "color": "#1D1E20",
          "fontFamily": "Inter_500Medium",
          "fontSize": 11,
          "marginBottom": 5,
        },
      ]
    }
  >
    shoes
  </Text>
  <Text
    style={
      Object {
        "color": "#1D1E20",
        "fontFamily": "Inter_600SemiBold",
        "fontSize": 13,
        "marginTop": 5,
      }
    }
  >
    $ 100
  </Text>
</View>
`;
